### PR TITLE
#161 : PowerPoint2007 : Read/Write SlideLayout 

### PR DESCRIPTION
**Pull request by Progi1984**

### CODE REVIEW

**Comment by vincentKool** [comment]
Looks like the CI problems are only one newline after namespace declaration and a newline at the end of a couple of files.

When making those adjustments, you could also change the comment in src/PhpPresentation/Shape/Placeholder.php (that file contains template comment from phpstorm)

**Comment by Progi1984** [comment]
@vincentKool I just fixed PHPCS errors and the comment header.

**Comment by Progi1984** [comment]
@vincentKool Have you some time for testing this patch ?

**Comment by Progi1984** [comment]
@vincentKool Have you some time for testing this patch ?

**Comment by vincentKool** [comment]
I can probably find some time for testing this weekend

**Comment by vincentKool** [comment]
Made a pull request in your Progi1984:issue161 branch for fixing the CI errors
 [Travis-CI](https://travis-ci.org/vincentKool/PHPPresentation/builds/141828272)

**Comment by ricovandevin** [comment]
I'd love to use this functionality. What is the status of it? Can I help testing in some way?

**Comment by Progi1984** [comment]
@ricovandevin Answer in #161 ;)

**Comment by Progi1984** [comment]
@vincentKool With the last version of the PR, i have an error on opening a Powerpoint 2007.

OpenXML SDK Validator says : 

screen shot 08-29-16 at 11 22 pm

An idea ?

**Comment by Progi1984** [comment]
@vincentKool I fix errors with always error on opening... shit !

File: src/PhpPresentation/Writer/PowerPoint2007/ContentTypes.php [hunk 1]
```diff
@@ -69,7 +69,7 @@ public function render()
             $oSlideMaster->setRelsIndex($idx + 1);
             $this->writeOverrideContentType($objWriter, '/ppt/slideMasters/slideMaster' . $oSlideMaster->getRelsIndex() . '.xml', 'application/vnd.openxmlformats-officedocument.presentationml.slideMaster+xml');
             $this->writeOverrideContentType($objWriter, '/ppt/theme/theme' . $oSlideMaster->getRelsIndex() . '.xml', 'application/vnd.openxmlformats-officedocument.theme+xml');
-            foreach ($oSlideMaster->getAllSlideLayouts() as &$oSlideLayout) {
```
Comment (Progi1984) [inline src/PhpPresentation/Writer/PowerPoint2007/ContentTypes.php]: Note for me : If we remove the reference, how the layoutNr & layoutId are updated ?

File: src/PhpPresentation/Reader/PowerPoint2007.php [hunk 2]
```diff
@@ -17,14 +17,23 @@
 
 namespace PhpOffice\PhpPresentation\Reader;
 
+/** This file will be edited to also read slideLayout things */
```
Comment (Progi1984) [inline src/PhpPresentation/Reader/PowerPoint2007.php]: Note for me : Comment to remove

File: src/PhpPresentation/Style/TextStyle.php [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    }
+
+    /**
+     * @return array
+     */
+    public function getOtherStyle()
+    {
+        return $this->otherStyle;
+    }
+//
```
Comment (Progi1984) [inline src/PhpPresentation/Style/TextStyle.php]: Note for me : Clean following lines...

File: src/PhpPresentation/Writer/PowerPoint2007/ContentTypes.php [hunk 4]
```diff
@@ -94,7 +86,7 @@ public function render()
         }
 
         // Add layoutpack content types
```
Comment (Progi1984) [inline src/PhpPresentation/Writer/PowerPoint2007/ContentTypes.php]: Note for me : clean following lines

File: src/PhpPresentation/Writer/PowerPoint2007/PptSlideLayouts.php [hunk 5]
```diff
@@ -2,29 +2,34 @@
...
-            $this->oZip->addFromString('ppt/slideLayouts/_rels/slideLayout' . $key . '.xml.rels', $this->writeSlideLayoutRelationships($key, $layout['masterid']));
-            $this->oZip->addFromString('ppt/slideLayouts/slideLayout' . $key . '.xml', utf8_encode($layout['body']));
+        foreach ($this->oPresentation->getAllMasterSlides() as $oSlideMaster) {
+            foreach ($oSlideMaster->getAllSlideLayouts() as $oSlideLayout) {
+                $this->oZip->addFromString('ppt/slideLayouts/_rels/slideLayout' . $oSlideLayout->layoutNr . '.xml.rels', $this->writeSlideLayoutRelationships($oSlideLayout->layoutNr, $oSlideMaster->getRelsIndex()));
+                $this->oZip->addFromString('ppt/slideLayouts/slideLayout' . $oSlideLayout->layoutNr . '.xml', $this->writeSlideLayout($oSlideLayout));
+            }
         }
 
+        /*
```
Comment (Progi1984) [inline src/PhpPresentation/Writer/PowerPoint2007/PptSlideLayouts.php]: Note for me : clean following lines

File: src/PhpPresentation/Writer/PowerPoint2007/PptSlideLayouts.php [hunk 6]
```diff
@@ -57,15 +59,120 @@ public function writeSlideLayoutRelationships($slideLayoutIndex, $masterId = 1)
         $this->writeRelationship($objWriter, 1, 'http://schemas.openxmlformats.org/officeDocument/2006/relationships/slideMaster', '../slideMasters/slideMaster' . $masterId . '.xml');
 
         // Other relationships
+        //@todo
```
Comment (Progi1984) [inline src/PhpPresentation/Writer/PowerPoint2007/PptSlideLayouts.php]: Note for me : :question:

File: src/PhpPresentation/Writer/PowerPoint2007/PptSlideMasters.php [hunk 7]
```diff
@@ -1,85 +1,229 @@
...
-                $this->oZip->addFromString($this->absoluteZipPath('ppt/slideMasters/' . $otherRelation['target']), $otherRelation['contents']);
-            }
+        foreach ($this->oPresentation->getAllMasterSlides() as $oMasterSlide) {
+            // Add the relations from the masterSlide to the ZIP file
+            $this->oZip->addFromString('ppt/slideMasters/_rels/slideMaster' . $oMasterSlide->getRelsIndex() . '.xml.rels', $this->writeSlideMasterRelationships($oMasterSlide));
+            // Add the information from the masterSlide to the ZIP file
+            $this->oZip->addFromString('ppt/slideMasters/slideMaster' . $oMasterSlide->getRelsIndex() . '.xml', $this->writeSlideMaster($oMasterSlide));
         }
-
+
```
Comment (Progi1984) [inline src/PhpPresentation/Writer/PowerPoint2007/PptSlideMasters.php]: Note for me : Clean following lines

File: src/PhpPresentation/Writer/PowerPoint2007/PptTheme.php [hunk 8]
```diff
@@ -13,21 +13,19 @@ class PptTheme extends AbstractDecoratorWriter
...
-        foreach ($masterSlides as $masterSlide) {
-            // Add themes to ZIP file
-            $this->getZip()->addFromString('ppt/theme/_rels/theme' . $masterSlide['masterid'] . '.xml.rels', $this->writeThemeRelationships($masterSlide['masterid']));
-            $this->getZip()->addFromString('ppt/theme/theme' . $masterSlide['masterid'] . '.xml', $this->writeTheme($masterSlide['masterid']));
+        foreach ($this->oPresentation->getAllMasterSlides() as $oMasterSlide) {
+            $this->getZip()->addFromString('ppt/theme/_rels/theme' . $oMasterSlide->getRelsIndex() . '.xml.rels', $this->writeThemeRelationships($oMasterSlide->getRelsIndex()));
+            $this->getZip()->addFromString('ppt/theme/theme' . $oMasterSlide->getRelsIndex() . '.xml', $this->writeTheme($oMasterSlide));
         }
 
+        /*
```
Comment (Progi1984) [inline src/PhpPresentation/Writer/PowerPoint2007/PptTheme.php]: Note for me : Clean following lines

File: src/PhpPresentation/Writer/PowerPoint2007/PptTheme.php [hunk 9]
```diff
@@ -75,13 +146,16 @@ public function writeThemeRelationships($masterId = 1)
         $objWriter->writeAttribute('xmlns', 'http://schemas.openxmlformats.org/package/2006/relationships');
 
         // Other relationships
+        //@todo
```
Comment (Progi1984) [inline src/PhpPresentation/Writer/PowerPoint2007/PptTheme.php]: Note for me : :question:

File: src/PhpPresentation/Writer/PowerPoint2007/PptTheme.php [hunk 10]
```diff
@@ -75,13 +146,16 @@ public function writeThemeRelationships($masterId = 1)
...
+        $masterId;
+        /*
```
Comment (Progi1984) [inline src/PhpPresentation/Writer/PowerPoint2007/PptTheme.php]: Note for me : Clean following lines

File: src/PhpPresentation/Reader/PowerPoint2007.php (hunk 2 above)
Comment (Progi1984) [inline src/PhpPresentation/Reader/PowerPoint2007.php]: Done

**Pull request merged**